### PR TITLE
Alert headers in email notifications [OTP-923]

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [na] The description lists all applicable issues this PR seeks to resolve
- [na] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR adds alert headers and improves alert formatting in trip email notifications.
Examples of improved renderings are shown in the screenshots below.

HTML email:
HTML email screenshot

Plain text email:
Plain-terxt email screenshot

### CODE REVIEW

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            return null;
+        }
+
+        TripMonitorAlertSubNotification newAlertsNotification = null;
+        TripMonitorAlertSubNotification resolvedAlertsNotification = null;
+
+        // If there are any unseen alerts, include list of these.
+        if (!unseenAlerts.isEmpty()) {
+            newAlertsNotification = new TripMonitorAlertSubNotification(
+                "New alerts found:",
```
Comment (miles-grant-ibigroup) [inline src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java:56]: Are we ok to leave this unlocalized for now?

**Review by miles-grant-ibigroup — APPROVED** [review]
Great code! I worry a little bit about localization, but I guess that's a future struggle

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java (hunk 1 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java:56]: I have not started on localization yet...

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        // Unseen alerts consists of all new alerts that we did not previously track.
+        HashSet<LocalizedAlert> unseenAlerts = new HashSet<>(newAlerts);
+        unseenAlerts.removeAll(previousAlerts);
+        // Resolved alerts consists of all previous alerts that no longer exist.
+        HashSet<LocalizedAlert> resolvedAlerts = new HashSet<>(previousAlerts);
+        resolvedAlerts.removeAll(newAlerts);
+        // If there is no change in alerts from previous check, no notification should be created.
+        if (unseenAlerts.isEmpty() && resolvedAlerts.isEmpty()) {
+            return null;
+        }
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java:48]: An identical code block to this exists in `TripMonitorNotification`. Can it be moved into it's own method?

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    /**
+     * Basic text formatting of the alert notification.
+     */
+    public static String getBody(
+        TripMonitorAlertSubNotification newAlertsNotification,
+        TripMonitorAlertSubNotification resolvedAlertsNotification
+    ) {
+        StringBuilder body = new StringBuilder();
+        if (newAlertsNotification != null) {
+            body.append(newAlertsNotification.toString());
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java]: `toString()` not needed.

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            body.append("\n");
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java]: Can be replace with `System.lineSeparator()`.

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+import java.util.Collection;
+
+/**
+ * Sub-notification for {@link TripMonitorAlertNotification}.
+ */
+public class TripMonitorAlertSubNotification extends TripMonitorNotification {
+    private final Collection<LocalizedAlert> alerts;
+
+    // Getter functions used by the HTML template renderer
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java]: Nit. Full stop. Also, is the new line between comment and method intentional?

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    public String toString() {
+        return toString(false);
+    }
+
+    public String toString(boolean resolved) {
+        StringBuilder result = new StringBuilder();
+        result.append(this.body);
+        result.append("\n");
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java]: Can be replace with System.lineSeparator().

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        for (LocalizedAlert alert : alerts) {
+            result.append("\n- ");
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java]: Can be replace with System.lineSeparator().

File: src/main/java/org/opentripplanner/middleware/otp/response/LocalizedAlert.java [hunk 8]
```diff
@@ -13,6 +16,28 @@ public class LocalizedAlert {
...
+
+    /**
+     * Line returns are not preserved if using the HTML email renderer,
+     * so we insert line returns as line-break (br) tags to match the itinerary-body UI.
+     */
+    @JsonIgnore
+    @BsonIgnore
+    public String getAlertDescriptionForHtml() {
+        return alertDescriptionText != null
+            ? alertDescriptionText.replace("\n", "<br/>\n")
```
Comment (br648) [inline src/main/java/org/opentripplanner/middleware/otp/response/LocalizedAlert.java]: Can be replace with System.lineSeparator().

**Review by br648 — COMMENTED** [review]
Some minor comments. Is it possible to update `CheckMonitoredTripTest` to include tests covering `TripMonitorAlertNotification` and `TripMonitorAlertSubNotification`?

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java (hunk 5 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java]: Yeah, there are multiple getters, so I didn't want to "attach" the comment to the first function.

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java (hunk 2 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java:48]: Good catch, that code in `TripMonitorNotification` ended up unused, so I deleted it from that file in https://github.com/ibi-group/otp-middleware/pull/204/commits/149f430fe7655b728a35cf8211c64bda231ac68e.

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java (hunk 3 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java]: Fixed in https://github.com/ibi-group/otp-middleware/pull/204/commits/5bed973a959c42bc530c5c5dfe647838f7f59d69;

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java (hunk 4 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertNotification.java]: Fixed in https://github.com/ibi-group/otp-middleware/pull/204/commits/5bed973a959c42bc530c5c5dfe647838f7f59d69;

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java (hunk 6 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java]: Fixed in https://github.com/ibi-group/otp-middleware/pull/204/commits/5bed973a959c42bc530c5c5dfe647838f7f59d69;

File: src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java (hunk 7 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/models/TripMonitorAlertSubNotification.java]: Fixed in https://github.com/ibi-group/otp-middleware/pull/204/commits/5bed973a959c42bc530c5c5dfe647838f7f59d69;

File: src/main/java/org/opentripplanner/middleware/otp/response/LocalizedAlert.java (hunk 8 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/middleware/otp/response/LocalizedAlert.java]: Fixed in https://github.com/ibi-group/otp-middleware/pull/204/commits/5bed973a959c42bc530c5c5dfe647838f7f59d69;

**Comment by binh-dam-ibigroup** [comment]
> Some minor comments. Is it possible to update `CheckMonitoredTripTest` to include tests covering `TripMonitorAlertNotification` and `TripMonitorAlertSubNotification`?

Yes, thank you, I added the tests in https://github.com/ibi-group/otp-middleware/pull/204/commits/28cfd9dd4d35b79a7dab6f5f445dbf756cf7425a and https://github.com/ibi-group/otp-middleware/pull/204/commits/ffbce7e22c061a734dea552aa98ff0b07e2ed341.

**Review by br648 — APPROVED** [review]
This looks good now. The unit tests are a good addition.

**Pull request merged**